### DISCUSSION TITLE
How to render pdf book with English and Arabic using either babel or polyglossia

### DISCUSSION THREAD

**Discussion by ghost** [quarto-dev/quarto-cli#6946] [Q&A]
### Description

I am struggling to render my English-language book with some Arabic in it to pdf. Could someone suggest a workable yaml code for me? In the _quarto.yml file for the book, I have tried the following polyglossia and babel codes, with the error messages below:

```
include-in-header:
      text: |
        \usepackage{polyglossia}
        \setdefaultlanguage{english}
        \setotherlanguage{arabic}
        \newfontfamily/arabicfont[Script=Arabic]{Amiri} 
```
Error message from xelatex compiler: compilation failed- package installation error
Package fontspec Error: The font "a" cannot be found.

Babel:

```
include-in-header:
      text: |
        \usepackage{babel}
        \babelfont[arabic]{rm}[Scale=1.7]{Amiri} 
```

Error message: LaTeX Error: Option clash for package babel.

**Comment by mcanouil** [comment]
Without a reproducible, simple, and complete example, it is going to be hard if not impossible to help.

Your question is mostly (if not exclusively related to the use of multiple languages in LaTeX).
Use `keep-tex: true` to debug your LaTeX intermediate file, once you know which part is causing the issue (likely "babel" as the error you showed), you can either use the `include-` key, modify the template or the partials to fix your LaTeX conflicts (see <https://quarto.org/docs/journals/templates.html#latex-partials>).

---


You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by ghost** [reply]
Thank you. My problem is that the document won't render at all (I'm using the Quarto renderer in VSCodium), so I don't have an example to show other than the yaml file for the book project. It renders in other formats - but I want it to work in pdf format, too.

**Reply by mcanouil** [reply]
As I said, use `keep-tex: true`, you'll have the intermediary LaTeX file that you can use to debug your LaTeX code.
Using the UI of whatever editor or directly `quarto render` in a terminal will work exactly the same with the above mentioned option in your YAML.

I don't believe the issue is specific to book format, so make a small document in another "project", using both english and arabic, try to render it.

How well do you know LaTeX?

**Comment by adamiturabi** [comment]
Here is how I got Quatro to work with a text that is mainly in English with Arabic divs and spans.

`example.qmd` file:

```
---
title: "Arabic-support Example"
keep-md: true
format:
  pdf:
    keep-tex: true
    pdf-engine: xelatex
    # babelfont for Arabic font needs to be in the `template-partial`
    # `before-title.tex`. If I put it in the `include-in-header` option, then it
    # gets put before `\usepackage{babel}` and doesn't work.
    template-partials:
      - before-title.tex
---

## Heading with inline Arabic [هَـٰذَا كَلَامٌ عَرَبِيٌّ.]{lang="ar"}

This is some example Arabic inline text [هَـٰذَا كَلَامٌ عَرَبِيٌّ.]{lang="ar"}.

This is an Arabic div:

::: {.otherlanguage data-latex="{arabic}" lang="ar"}
هذا كلام عربي طويل. أريد أن أكتب حتى يبلغ الكلام سطرين. أستعمل برنامج قواطرو لإنتاج الملف الخارجي. هو برنامج جيد قد خلف البرنامج بكداؤن الذي كنت أستعمله من قبل.
:::

This is some English text after it.
```

`before-title.tex`:

```
\babelfont[arabic]{rm}[Language=Default]{DejaVu Sans}
\babelfont[arabic]{sf}[Language=Default]{DejaVu Sans}
```

Command:

```
quarto render example.qmd
```

This is a screenshot of the rendered output PDF:
<img width="653" alt="image" src="https://github.com/user-attachments/assets/467cd7b9-e600-4172-b401-7dd066623cfc">

Hope this helps.

**Reply by mcanouil** [reply]
Thanks for sharing!

**Reply by adamiturabi** [reply]
By the way, I found that for HTML output it is better for the span to have the attribute `dir="rtl"` in addition to `lang="ar"`, thus: `[كلام عربي]{dir="rtl" lang="ar"}`.

But for Latex output, `dir="rtl"` clashes with the Xelatex PDF engine.

I'm managing this difference using a pandoc filter so that I only type `[كلام عربي]{.ar}` in the source `.qmd` file. Here is my filter:

```
function Span (el)
  if el.classes:includes 'ar' then
    text = pandoc.utils.stringify(el)
    contents = {pandoc.Str(text)}
    if FORMAT:match 'latex' then
      -- no dir needed for babel and throws error if it sees dir attribute. was previously needed for polyglossia
      return pandoc.Span(contents, {class='regartext', lang='ar'})
    elseif FORMAT:match 'html' then
      -- dir needed for html otherwise punctuation gets messed up
      return pandoc.Span(contents, {class='regartext', lang='ar', dir='rtl'})
    end
  end
end
```

I have a MWE in this repo: https://github.com/adamiturabi/quarto-arabic-mwe

**Reply by mcanouil** [reply]
Smart and thanks for the additional details.
